### PR TITLE
Refactor to avoid suppressing CA2208

### PR DESCRIPTION
**Pull request by donnie-msft**
## Bug

Fixes: https://github.com/nuget/client.engineering/issues/642

## Fix

Removed all CA2208 suppressions. Commits are per-file (except for "Missed random CA2208s").

- Created a new `PropertyCannotBeNull` String Resources so the messages can explain that property {0} is null. I copied this from `ClientCertificatesCommand` resx file.

- I used some `ArgumentOutOfRangeExceptions` in a few spots for things like Enum `default` blocks. I'm not sure this is completely an accurate exception type here. On the other hand, I'm fairly sure in most instances, they aren't user-level errors, anyway. Using this exception type makes it explicit that there's a value problem, without having to use Strings to explain it.

- SemanticVersion is an example of an outlier. Here I don't think a user will ever see this, so creating a RESX file seems overkill. `ArgumentException("Argument must be of type 'SemanticVersion'.", nameof(obj));`

- VSMachineWideSettings.cs - I switched this to a `NullReferenceException` since an ArgumentException in a ctor doesn't make sense to me, and there's no parameter to provide.

## Testing/Validation

Tests Added: Yes
Validation:  A few unit tests were added to prove that the exception is thrown, and contains the property name in the message.

### CODE REVIEW

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs [hunk 1]
```diff
@@ -2805,11 +2817,6 @@ var projectUniqueNamesForBuildIntToUpdate
 
                 if (packageIdentity != null)
                 {
-                    if (primarySources == null || primarySources.Count == 0)
-                    {
```
Comment (dominoFire) [inline src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:2809]: Do you also need to create a string resource for line 2828 ?

File: src/NuGet.Clients/NuGet.VisualStudio/LegacyTypes/SemanticVersion.cs [hunk 2]
```diff
@@ -210,7 +210,7 @@ public int CompareTo(object obj)
             SemanticVersion other = obj as SemanticVersion;
             if (other == null)
             {
-                throw new ArgumentException("obj");
+                throw new ArgumentException("Argument must be of type 'SemanticVersion'.", nameof(obj));
```
Comment (dominoFire) [inline src/NuGet.Clients/NuGet.VisualStudio/LegacyTypes/SemanticVersion.cs]: I feel we need a string resource for this message:

```suggestion
                throw new ArgumentException(string.Format(Strings.ArgumentTypeExceptionMessage, nameof(SemanticVersion)), nameof(obj));
```

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UpgradeLogger.cs [hunk 3]
```diff
@@ -80,7 +80,7 @@ internal UpgradeLogger(string reportName, string backupPath)
         {
             if (string.IsNullOrEmpty(reportName))
             {
-                throw new ArgumentException(nameof(reportName));
+                throw new ArgumentNullException(nameof(reportName));
```
Comment (dominoFire) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UpgradeLogger.cs]: Why `ArgumentNullException`? See line 81.

File: src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs [hunk 4]
```diff
@@ -848,7 +848,7 @@ private static void AddProperty(List<string> args, string property, string value
         {
             if (string.IsNullOrEmpty(value))
             {
-                throw new ArgumentException(nameof(value));
+                throw new ArgumentNullException(nameof(value));
```
Comment (dominoFire) [inline src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs]: Why `ArgumentNullException`? See line 849.

**Review by dominoFire — COMMENTED** [review]
Looks good. Some comments left.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/VSMachineWideSettings.cs [hunk 5]
```diff
@@ -20,7 +21,8 @@ public VsMachineWideSettings()
         {
             if (_asyncServiceProvider == null)
             {
-                throw new ArgumentNullException(nameof(_asyncServiceProvider));
+                throw new NullReferenceException(
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/VSMachineWideSettings.cs]: This one makes me so uneasy. 

NullReferenceException seems like an uncontrolled exception. 

Maybe we can refactor VsMachineWideSettings to have 2 constructors, and the internal constructor can have the argument check.

See https://github.com/NuGet/NuGet.Client/blob/e9722d69400a57d2c0508bbce9e9591fcad550f1/src/NuGet.Clients/NuGet.Console/OutputConsole/OutputConsoleProvider.cs#L29-L40 for example

**Review by nkolev92 — COMMENTED** [review]
A quick look doesn't raise any concerns. 

I have one suggestion, it's great beyond that.

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs (hunk 1 above)
Comment (donnie-msft) [reply src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:2809]: After some offline debate.... yes. I added in new commit.

File: src/NuGet.Clients/NuGet.VisualStudio/LegacyTypes/SemanticVersion.cs (hunk 2 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.VisualStudio/LegacyTypes/SemanticVersion.cs]: Added RESX file... do I need to do anything special for new files?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UpgradeLogger.cs (hunk 3 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UpgradeLogger.cs]: Fixed

File: src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs (hunk 4 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs]: Fixed

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/VSMachineWideSettings.cs (hunk 5 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/VSMachineWideSettings.cs]: Yeah, I called out the NRE in the PR description as being weird. Though I don't know an alternative exception type to use. I do like this pattern, see if you like the latest commit.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/VerifyCommand.cs [hunk 6]
```diff
@@ -36,7 +36,7 @@ public override Task ExecuteCommandAsync()
 
             if (string.IsNullOrEmpty(PackagePath))
             {
-                throw new ArgumentNullException(nameof(PackagePath));
+                throw new ArgumentException(nameof(PackagePath));
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/VerifyCommand.cs:39]: This is technically not correct use of the exception... the parameter here is actually "message", so there's no argument given. But this isn't an argument anyway....?! I think it's probably overkill to refactor this.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/VSMachineWideSettings.cs (hunk 5 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/VSMachineWideSettings.cs]: yep, looks good to me.

**Review by nkolev92 — APPROVED** [review]

File: src/NuGet.Clients/NuGet.VisualStudio/LegacyTypes/SemanticVersion.cs (hunk 2 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.VisualStudio/LegacyTypes/SemanticVersion.cs]: NuGet.VisualStudio.dll is already picked up for localization. No further action is required to add the new RESX file. I will keep an eye on it.

**Review by dominoFire — APPROVED** [review]

**Review by erdembayar — APPROVED** [review]

File: src/NuGet.Clients/NuGet.CommandLine/Commands/VerifyCommand.cs (hunk 6 above)
Comment (kartheekp-ms) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/VerifyCommand.cs:39]: I think this case falls under `CommandException` as `PackagePath` is an argument to `nuget verify` command. Inspired from below code.
https://github.com/NuGet/NuGet.Client/blob/a52c0addf437f4e358f786acadd4c91a2c80f9b6/src/NuGet.Clients/NuGet.CommandLine/Commands/AddCommand.cs#L31-L35

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Strings.Designer.cs [hunk 7]
```diff
@@ -411,6 +411,15 @@ public class Strings {
...
         
+        /// <summary>
+        ///   Looks up a localized string similar to &apos;{0}&apos; cannot be null..
+        /// </summary>
+        public static string PropertyCannotBeNull {
+            get {
+                return ResourceManager.GetString("PropertyCannotBeNull", resourceCulture);
+            }
+        }
+
```
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Strings.Designer.cs:422]: `NuGet.PackageManagement.VisualStudio` project already has an entry with key `Argument_Cannot_Be_Null_Or_Empty`. How about reusing the same entry instead of creating a new one?
https://github.com/NuGet/NuGet.Client/blob/a52c0addf437f4e358f786acadd4c91a2c80f9b6/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Strings.Designer.cs#L111-L114

**Review by kartheekp-ms — COMMENTED** [review]
I am still going through these changes. Left couple of comment to start with.

File: src/NuGet.Clients/NuGet.VisualStudio/LegacyTypes/SemanticVersion.cs [hunk 8]
```diff
@@ -134,15 +135,15 @@ private static string[] SplitAndPadVersionString(string version)
...
+            if (string.IsNullOrEmpty(version))
             {
                 throw new ArgumentException("Value cannot be null or an empty string.", "version");
             }
 
             SemanticVersion semVer;
             if (!TryParse(version, out semVer))
             {
-                throw new ArgumentException(String.Format(CultureInfo.CurrentCulture, "'{0}' is not a valid version string.", version), "version");
+                throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, "'{0}' is not a valid version string.", version), "version");
```
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.VisualStudio/LegacyTypes/SemanticVersion.cs:146]: This comment is out of scope for this PR. Hard-coded strings will not be localized. I think we should add an entry in resource file. @dominoFire thoughts?

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Strings.Designer.cs (hunk 7 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Strings.Designer.cs:422]: Yes, that one doesn't take a parameter, and isn't accurate anyway. The way I'm using it is that "a property on an argument" is null. Not the argument itself.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/VerifyCommand.cs (hunk 6 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/VerifyCommand.cs:39]: Are we comfortable changing the exception type?

Comment (kartheekp-ms) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/VerifyCommand.cs:39]: I guess `CommandException` or `ArgumentException` really don't matter much as long as we have some user friendly message. Can we add a resource entry for this case and use that for `message` parameter for `ArgumentException` constructor? This should address your [first ](https://github.com/NuGet/NuGet.Client/pull/3799/files#r544566747) comment on this thread.
https://github.com/NuGet/NuGet.Client/blob/a52c0addf437f4e358f786acadd4c91a2c80f9b6/src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx#L5493-L5497

File: src/NuGet.Clients/NuGet.VisualStudio/LegacyTypes/SemanticVersion.cs (hunk 8 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.VisualStudio/LegacyTypes/SemanticVersion.cs:146]: I think it's a good idea to put strings in string resources

Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.VisualStudio/LegacyTypes/SemanticVersion.cs:146]: Yep. I have seen this error message A LOT in VS Feedback, etc. I think this is one of the most common error messages.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/VerifyCommand.cs (hunk 6 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/VerifyCommand.cs:39]: You have some evidence that CommandException is more of a standard. How about we merge this PR, and have a separate issue to switch that one?

Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/VerifyCommand.cs:39]: https://github.com/NuGet/Home/issues/10391

File: src/NuGet.Clients/NuGet.VisualStudio/LegacyTypes/SemanticVersion.cs (hunk 8 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.VisualStudio/LegacyTypes/SemanticVersion.cs:146]: https://github.com/NuGet/Home/issues/10392
